src/main/java/games/strategy/engine/lobby/server/login/RsaAuthenticator.java

```diff
@@ -0,0 +1,183 @@
+    final PrivateKey privateKey = rsaKeyMap.get(publicKey);
+    if (privateKey == null) {
+      return "Login timeout, try again!";
+    }
+    try {
+      final Cipher cipher = Cipher.getInstance(RSA_ECB_OAEPP);
+      cipher.init(Cipher.DECRYPT_MODE, privateKey);
+      return function
+          .apply(new String(cipher.doFinal(Base64.getDecoder().decode(encryptedPassword)), StandardCharsets.UTF_8));
+    } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+      throw new IllegalStateException(e);
+    } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException e) {
+      return e.getMessage();
+    } finally {
+      rsaKeyMap.remove(publicKey);
+    }
+  }
+
+  /**
+   * Attempts to decrypt the given password using the challenge and response parameters.
+   */
+  public static String decryptPassword(final Map<String, String> challenge, final Map<String, String> response,
+      final Function<String, String> successFullDecryptionAction) {
+    return decryptPassword(response.get(ENCRYPTED_PASSWORD_KEY), challenge.get(RSA_PUBLIC_KEY),
+        successFullDecryptionAction);
+  }
+
+
+  @VisibleForTesting
+  static String encryptPassword(final String publicKeyString, final String password) {
+    try {
+      final PublicKey publicKey =
+          KeyFactory.getInstance(RSA)
+              .generatePublic(new X509EncodedKeySpec(Base64.getDecoder().decode(publicKeyString)));
+      final Cipher cipher = Cipher.getInstance(RSA_ECB_OAEPP);
+      cipher.init(Cipher.ENCRYPT_MODE, publicKey);
+      return Base64.getEncoder().encodeToString(cipher.doFinal(Util.sha512(password).getBytes(StandardCharsets.UTF_8)));
+    } catch (GeneralSecurityException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * This method adds the encrypted password (using the specified public key and password)
+   * to the specified response map.
+   */
+  public static void appendEncryptedPassword(
+      final Map<String, String> response,
+      final Map<String, String> challenge,
+      final String password) {
+    response.put(ENCRYPTED_PASSWORD_KEY, encryptPassword(challenge.get(RSA_PUBLIC_KEY), password));
+  }
+  
+  @VisibleForTesting
+  static void setTimeout(final long amount, final TimeUnit timeUnit) {
+    rsaKeyMap.setTimeout(amount, timeUnit);
+  }
+
+  private static class TimeoutKeyMap extends HashMap<String, PrivateKey> {
+    private static final long serialVersionUID = 3788873489149542052L;
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+    private final Map<String, ScheduledFuture<?>> scheduledTaks = new HashMap<>();
+    private long amount;
+    private TimeUnit timeUnit;
+
+    private TimeoutKeyMap() {
+      setTimeout(10, TimeUnit.MINUTES);
+    }
+
+    @Override
+    public PrivateKey put(final String string, final PrivateKey key) {
+      scheduledTaks.put(string, scheduler.schedule(() -> {
+        super.remove(string);
```

Comment:
Instance state is being accessed from multiple threads, but this class is not thread safe.